Giant Board support (SAMA5D27 board)

adamfeuer opened the pull request:
## Summary
- Giant Board support (SAMA5D27 board)
  - see https://groboards.com/giant-board/
- tested with CDC/ECM USB over Ethernet, SDMMC, and nsh

## Impact
- adds the following configs 
  - `giant-board:nsh`
  - `giant-board:sdmmc-net-nsh`
  - `giant-board:netnsh`
  - `giant-board:sdmmcnsh`

## Testing
- all configs were manually tested

## Detail
- there are a few nxstyle errors with no obvious way to correct them

ghn-certi commented on boards/Kconfig:
```diff
@@ -1364,13 +1364,21 @@ config ARCH_BOARD_SABRE_6QUAD
 	---help---
-		The port of NuttX to the Atmel SAMA5D2-XULT development board
+	    This options selects support for NuttX on the Groboards Giant Board,
+	    featuring the SAMA5D27D1G SIP with 128KB DDR3 RAM.
```
According to the documentation, the SiP provides 128MB DDR2 RAM.

adamfeuer replied:
@ghn-certi Will fix.

ghn-certi commented on boards/Kconfig:
```diff
@@ -1364,13 +1364,21 @@ config ARCH_BOARD_SABRE_6QUAD
 	---help---
-		The port of NuttX to the Atmel SAMA5D2-XULT development board
+	    This options selects support for NuttX on the Groboards Giant Board,
+	    featuring the SAMA5D27D1G SIP with 128KB DDR3 RAM.
```
I tried to google for the mentioned PN (SAMA5D27D1G) and it returned zero occurrences.
The correct PN for the SiP is ATSAMA5D27C-D1G.

adamfeuer replied:
@ghn-certi Will fix.

ghn-certi commented on boards/Kconfig:
```diff
@@ -1364,13 +1364,21 @@ config ARCH_BOARD_SABRE_6QUAD
 	---help---
-		The port of NuttX to the Atmel SAMA5D2-XULT development board
+	    This options selects support for NuttX on the Groboards Giant Board,
+	    featuring the SAMA5D27C-D1G SIP with 128KB DDR2 RAM.
```
```suggestion
	    featuring the SAMA5D27C-D1G SIP with 128MB DDR2 RAM.
```

ghn-certi commented on boards/arm/sama5/giant-board/include/board.h:
```diff
@@ -0,0 +1,316 @@
+/****************************************************************************
+ * boards/arm/sama5/giant-board/include/board.h
+ *
+ *   Copyright (C) 2015 Gregory Nutt. All rights reserved.
```
Since this is a newly created file, shouldn't you put the Apache copyright header?

adamfeuer replied:
@ghn-certi It's ported from SAMA5D2-XULT, but probably should have an ASF header. Will fix.

ghn-certi commented on boards/arm/sama5/giant-board/include/board.h:
```diff
@@ -0,0 +1,316 @@
+ * That bootloader had to have already configured the PLL and SDRAM for
+ * proper operation.
+ *
+ * In this case, we don not reconfigure the clocking.
```
```suggestion
 * In this case, we do not reconfigure the clocking.
```

ghn-certi commented on boards/arm/sama5/giant-board/include/board.h:
```diff
@@ -0,0 +1,316 @@
+
+#elif defined(CONFIG_GIANT_BOARD_492MHZ)
+
+/* This is the configuration results in a CPU clock of 498MHz.
```
There's something wrong in this sentence. See if my suggestion is what you meant to document.
```suggestion
/* This configuration results in a CPU clock of 492MHz.
```

ghn-certi commented on boards/arm/sama5/giant-board/include/board_492mhz.h:
```diff
@@ -0,0 +1,211 @@
+/****************************************************************************
+ * boards/arm/sama5/sama5d2-xult/include/board_498mhz.h
```
```suggestion
 * boards/arm/sama5/giant-board/include/board_492mhz.h
```

ghn-certi commented on boards/arm/sama5/giant-board/include/board_492mhz.h:
```diff
@@ -0,0 +1,211 @@
+ *
+ ****************************************************************************/
+
+#ifndef __BOARDS_ARM_SAMA5_GIANT_BOARD_INCLUDE_BOARD_498MHZ_H
+#define __BOARDS_ARM_SAMA5_GIANT_BOARD_INCLUDE_BOARD_498MHZ_H
```
```suggestion
#ifndef __BOARDS_ARM_SAMA5_GIANT_BOARD_INCLUDE_BOARD_492MHZ_H
#define __BOARDS_ARM_SAMA5_GIANT_BOARD_INCLUDE_BOARD_492MHZ_H
```

ghn-certi commented on boards/arm/sama5/giant-board/include/board_492mhz.h:
```diff
@@ -0,0 +1,211 @@
+/* After power-on reset, the SAMA5 device is running on a 12MHz internal RC.
+ * These definitions will configure operational clocking.
+ *
+ * This is the configuration results in a CPU clock of 498MHz:
```
```suggestion
 * This configuration results in a CPU clock of 492MHz:
```

ghn-certi commented on boards/arm/sama5/giant-board/include/board_492mhz.h:
```diff
@@ -0,0 +1,211 @@
+
+/* PLLA configuration.
+ *
+ *   Multipler = 20+1: PLLACK = 21 * 24MHz = 498MHz
```
This documentation is not coherent with the one on line 42.
Since the PLLA is already documented above, is it really needed to repeat here?
```suggestion
 *   Multiplier = 40+1: PLLACK = 41 * 24MHz = 984MHz
```

adamfeuer replied:
@ghn-certi I just copied this file from the SAMA5D2-XULT one, was trying to stay true to what it was doing. I will take a closer look at this.

ghn-certi commented on boards/arm/sama5/giant-board/include/board_memorymap.h:
```diff
@@ -0,0 +1,74 @@
+/****************************************************************************
+ * boards/arm/sama5/sama5d2-xult/include/board_memorymap.h
```
```suggestion
 * boards/arm/sama5/giant-board/include/board_memorymap.h
```

ghn-certi commented on boards/arm/sama5/giant-board/include/board_memorymap.h:
```diff
@@ -0,0 +1,74 @@
+/****************************************************************************
+ * boards/arm/sama5/sama5d2-xult/include/board_memorymap.h
+ *
+ *   Copyright (C) 2015 Gregory Nutt. All rights reserved.
```
Apache header

adamfeuer replied:
@ghn-certi Will fix.

patacongo commented on boards/Kconfig:
```diff
@@ -1364,13 +1364,21 @@ config ARCH_BOARD_SABRE_6QUAD
+        select ARCH_HAVE_IRQBUTTONS
+---help---
+The port of NuttX to the Atmel SAMA5D2-XULT development board
+
```
The above does not follow the indentation rules that we use in Kconfig files.  Need a TAB before bool and depends and ---help--- and two TABs before "The port of NuttX"

adamfeuer replied:
@patacongo Will fix.

patacongo commented on boards/arm/sama5/giant-board/Kconfig:
```diff
@@ -0,0 +1,126 @@
+    select ARCH_HAVE_SDIO
+    select SAMA5_SDMMC_DMA
+    ---help---
+    Enable SD Card interface SDMMC0. Selects SAMA5_SDMMC SAMA5_SDMMC0 SAMA5_SDMMC_DMA SDIO_DMA SCHED_WORKQUEUE SCHED_HPWORK SDIO_BLOCKSETUP
```
Please divide this long line.  All comments are indented by two TABs.

patacongo replied:
The two TAB comment applies everywhere.

adamfeuer replied:
@patacongo Thanks, will fix.

patacongo commented on boards/arm/sama5/giant-board/helpers/netusb-up.sh:
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+set -x
```
All Bash scripts need and Apache 2.0 license on them

patacongo commented on boards/arm/sama5/giant-board/include/board.h:
```diff
@@ -0,0 +1,316 @@
+
+#define GB_PIO_SCL        PIO_TWI0_D_2          /* SDA / PB31 */
+#define GB_PIO_PB31       PIO_TWI0_D_2          /* SDA / PB31 */
+
```
The above definitions are all internal to the board implementation and must NOT be exposed in the board.h header file.  Please move to the giant-board.h header file.

adamfeuer replied:
@patacongo will fix.

patacongo commented on boards/arm/sama5/giant-board/scripts/Make.defs:
```diff
@@ -0,0 +1,103 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+############################################################################
```
Please convert this to an Apache 2.0 header.

adamfeuer replied:
@patacongo will fix.

patacongo commented on boards/arm/sama5/giant-board/scripts/dramboot.ld:
```diff
@@ -0,0 +1,138 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
```
Please convert this to an Apache 2.0 header

adamfeuer replied:
@patacongo will fix.

patacongo commented on boards/arm/sama5/giant-board/scripts/dramboot.ld:
```diff
@@ -0,0 +1,138 @@
+ ****************************************************************************/
+
+/* The SAMA5D4 has 128 KB of ISRAM beginning at virtual address 0x0020:0000.
+ * This memory configuration, however, loads into the 256MB SDRAM on board
```
```suggestion
 * This memory configuration, however, loads into the 128MB SDRAM on board
```

patacongo commented on boards/arm/sama5/giant-board/scripts/dramboot.ld:
```diff
@@ -0,0 +1,138 @@
+MEMORY
+{
+    isram (W!RX) : ORIGIN = 0x00200000, LENGTH = 128K - 16K
+    sdram (W!RX) : ORIGIN = 0x20000000, LENGTH = 256M
```
```suggestion
    sdram (W!RX) : ORIGIN = 0x20000000, LENGTH = 128M
```

patacongo commented on boards/arm/sama5/giant-board/scripts/gnu-elf.ld:
```diff
@@ -0,0 +1,139 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
```
Convert to Apache 2.0 header

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/scripts/isram.ld:
```diff
@@ -0,0 +1,134 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
```
Convert to Apache 2.0 header

adamfeuer replied:
@patacongo will fix.

patacongo commented on boards/arm/sama5/giant-board/scripts/dramboot.ld:
```diff
@@ -0,0 +1,138 @@
+ *
+ ****************************************************************************/
+
+/* The SAMA5D4 has 128 KB of ISRAM beginning at virtual address 0x0020:0000.
```
```suggestion
/* The SAMA5D27 has 128 KB of ISRAM beginning at virtual address 0x0020:0000.
```
Use full part name (I forget the rest of the part name

adamfeuer replied:
@patacongo will fix – it's SAMA5D2C-D1G.

patacongo commented on boards/arm/sama5/giant-board/scripts/uboot.ld:
```diff
@@ -0,0 +1,139 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
```
Convert to an Apache 2.0 header

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/scripts/uboot.ld:
```diff
@@ -0,0 +1,139 @@
+ *
+ ****************************************************************************/
+
+/* The SAMA5D4 has 128 KB of ISRAM beginning at virtual address 0x0020:0000.
```
```suggestion
/* The SAMA5D27 has 128 KB of ISRAM beginning at virtual address 0x0020:0000.
```
Use the full name for the part

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/scripts/uboot.ld:
```diff
@@ -0,0 +1,139 @@
+
+/* The SAMA5D4 has 128 KB of ISRAM beginning at virtual address 0x0020:0000.
+ * This memory configuration, however, loads into the 256MB SDRAM on board
+ * the SAMA5D4-EK which lies at 0x2000:0000.  An offset 0x00008000 is
```
```suggestion
 * the Giant Board which lies at 0x2000:0000.  An offset 0x00008000 is
```

patacongo commented on boards/arm/sama5/giant-board/scripts/uboot.ld:
```diff
@@ -0,0 +1,139 @@
+ ****************************************************************************/
+
+/* The SAMA5D4 has 128 KB of ISRAM beginning at virtual address 0x0020:0000.
+ * This memory configuration, however, loads into the 256MB SDRAM on board
```
```suggestion
 * This memory configuration, however, loads into the 128MB SDRAM on board
```

patacongo commented on boards/arm/sama5/giant-board/scripts/uboot.ld:
```diff
@@ -0,0 +1,139 @@
+MEMORY
+{
+    isram (W!RX) : ORIGIN = 0x00200000, LENGTH = 128K - 16K
+    sdram (W!RX) : ORIGIN = 0x20008000, LENGTH = 256M - 32K
```
```suggestion
    sdram (W!RX) : ORIGIN = 0x20008000, LENGTH = 128M - 32K
```

patacongo commented on boards/arm/sama5/giant-board/src/Makefile:
```diff
@@ -0,0 +1,146 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+############################################################################
```
Convert to an Apache 2.0 header

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,349 @@
+/****************************************************************************
+ * Public data
+ ****************************************************************************/
+
```
You can remove empty section headers

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_appinit.c:
```diff
@@ -0,0 +1,80 @@
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
```
You can remove empty section headers

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_at25.c:
```diff
@@ -0,0 +1,131 @@
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
```
You can remove empty section headers

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_at25.c:
```diff
@@ -0,0 +1,131 @@
+        }
+
+#if defined(CONFIG_GIANTBOARD_AT25_FTL)
+
```
Remove this blank line.  Customarily, there is no space between conditional logic and comments.  Not sure if that is part of the coding standard or not.

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_at25.c:
```diff
@@ -0,0 +1,131 @@
+        }
+
+#elif defined(CONFIG_GIANTBOARD_AT25_NXFFS)
+
```
Ditto

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_boot.c:
```diff
@@ -0,0 +1,98 @@
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
```
You can remove empty section headers

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  i2c = sam_i2cbus_initialize(bus);
+  if (i2c == NULL)
+    {
+      _err("ERROR: Failed to get I2C%d interface\n", bus);
```
```suggestion
      i2cerr("ERROR: Failed to get I2C%d interface\n", bus);
```
Please do not use _err().  That is intended only for internal use within debug.h.  Please change to

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+      ret = i2c_register(i2c, bus);
+      if (ret < 0)
+        {
+          _err("ERROR: Failed to register I2C%d driver: %d\n", bus, ret);
```
```suggestion
          i2cerr("ERROR: Failed to register I2C%d driver: %d\n", bus, ret);
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  sdmmc0 = sam_sdmmc_sdio_initialize(SDMMC0_SLOTNO);
+  if (!sdmmc0)
+    {
+      syslog(LOG_ERR, "ERROR: Failed to initialize SD/MMC\n");
```
mcerr() would be preferred over syslog().

adamfeuer replied:
@patacongo will change this to `mcerr()`. This came from the SAMA5D2-XULT code, I failed to update it, sorry. 🙂

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+
+  if (ret < 0)
+    {
+      _err("ERROR: Failed to mount %s: %d\n",
```
```suggestion
      mcerr("ERROR: Failed to mount %s: %d\n",
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+
+  if (ret < 0)
+    {
+      _err("ERROR: Failed to mount %s: %d\n",
```
```suggestion
      mcerr("ERROR: Failed to mount %s: %d\n",
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+                         CONFIG_GIANT_BOARD_ROMFS_ROMDISK_SECTSIZE);
+  if (ret < 0)
+    {
+      _err("ERROR: romdisk_register failed: %d\n", -ret);
```
```suggestion
      serr("ERROR: romdisk_register failed: %d\n", -ret);
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+                  "romfs", MS_RDONLY, NULL);
+      if (ret < 0)
+        {
+          _err("ERROR: mount(%s,%s,romfs) failed: %d\n",
```
```suggestion
          fserr("ERROR: mount(%s,%s,romfs) failed: %d\n",
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = sam_usbhost_initialize();
+  if (ret != OK)
+    {
+        _err("ERROR: Failed to initialize USB host: %d\n", ret);
```
```suggestion
        usberr("ERROR: Failed to initialize USB host: %d\n", ret);
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = usbmonitor_start();
+  if (ret != OK)
+    {
+        _err("ERROR: Failed to start the USB monitor: %d\n", ret);
```
```suggestion
        usberr("ERROR: Failed to start the USB monitor: %d\n", ret);
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = sam_tsc_setup(0);
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: sam_tsc_setup failed: %d\n", ret);
```
```suggestion
      tmrerr(LOG_ERR, "ERROR: sam_tsc_setup failed: %d\n", ret);
```

adamfeuer replied:
@patacongo will fix. (This came from SAMA5D2-XULT.)

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = sam_pwm_setup();
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: sam_pwm_setup() failed: %d\n", ret);
```
```suggestion
      pwmerr("ERROR: sam_pwm_setup() failed: %d\n", ret);
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = sam_adc_setup();
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: sam_adc_setup failed: %d\n", ret);
```
```suggestion
      aerr("ERROR: sam_adc_setup failed: %d\n", ret);
```

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = sam_wm8904_initialize(0);
+  if (ret != OK)
+    {
+      _err("ERROR: Failed to initialize WM8904 audio: %d\n", ret);
```
```suggestion
      auderr("ERROR: Failed to initialize WM8904 audio: %d\n", ret);
```

adamfeuer replied:
@patacongo will fix.

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = sam_audio_null_initialize(0);
+  if (ret != OK)
+    {
+      _err("ERROR: Failed to initialize the NULL audio device: %d\n", ret);
```
```suggestion
      auderr("ERROR: Failed to initialize the NULL audio device: %d\n", ret);
```

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+  ret = mount(NULL, SAMA5_PROCFS_MOUNTPOINT, "procfs", 0, NULL);
+  if (ret < 0)
+    {
+      _err("ERROR: Failed to mount procfs at %s: %d\n",
```
```suggestion
      fserr("ERROR: Failed to mount procfs at %s: %d\n",
```

adamfeuer replied:
@patacongo will fix.

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,401 @@
+    ret = cdcecm_initialize(0, NULL);
+  if (ret < 0)
+    {
+      _err("ERROR: cdcecm_initialize() failed: %d\n", ret);
```
```suggestion
      nerr("ERROR: cdcecm_initialize() failed: %d\n", ret);
```

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_buttons.c:
```diff
@@ -0,0 +1,139 @@
+/****************************************************************************
```
I think you need to delete this file.  The giant board has not buttons, does it?

adamfeuer replied:
@patacongo Ok, will delete. I left it because if you do add buttons yourself, you need some code to handle them; but I get that the user should go look for the code themselves and add it to their own application.

patacongo commented on boards/Kconfig:
```diff
@@ -1364,13 +1364,21 @@ config ARCH_BOARD_SABRE_6QUAD
+bool "Atmel SAMA5D2 Xplained Ultra development board"
+depends on ARCH_CHIP_ATSAMA5D27
+        select ARCH_HAVE_LEDS
+        select ARCH_HAVE_BUTTONS
```
Remove.  The Giant board does not have any buttons that are accessible by the user, does it?  I might be wrong.

patacongo commented on boards/Kconfig:
```diff
@@ -1364,13 +1364,21 @@ config ARCH_BOARD_SABRE_6QUAD
+depends on ARCH_CHIP_ATSAMA5D27
+        select ARCH_HAVE_LEDS
+        select ARCH_HAVE_BUTTONS
+        select ARCH_HAVE_IRQBUTTONS
```
Remove for the same reason.

patacongo commented on boards/arm/sama5/giant-board/configs/netnsh/defconfig:
```diff
@@ -0,0 +1,119 @@
+CONFIG_ARCH_CHIP_SAMA5=y
+CONFIG_ARCH_CHIP_SAMA5D2=y
+CONFIG_ARCH_INTERRUPTSTACK=2048
+CONFIG_ARCH_IRQBUTTONS=y
```
Ditto

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/configs/nsh/defconfig:
```diff
@@ -0,0 +1,80 @@
+CONFIG_ARCH="arm"
+CONFIG_ARCH_BOARD="sama5d2-xult"
+CONFIG_ARCH_BOARD_GIANT_BOARD=y
+CONFIG_ARCH_BUTTONS=y
```
See comments about.  Please refresh this configuration.  Remove button selections.

If the configuration is not tested, please remove it from the PR

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/configs/sdmmcnsh/defconfig:
```diff
@@ -0,0 +1,143 @@
+CONFIG_ARCH="arm"
+CONFIG_ARCH_BOARD="giant-board"
+CONFIG_ARCH_BOARD_GIANT_BOARD=y
+CONFIG_ARCH_BUTTONS=y
```
Refresh configuration.  There should be not button settings.

Did you test this configuration.  If not, please remove it.

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/configs/sdmmcnsh/defconfig:
```diff
@@ -0,0 +1,143 @@
+CONFIG_ARCH_CHIP_SAMA5=y
+CONFIG_ARCH_CHIP_SAMA5D2=y
+CONFIG_ARCH_INTERRUPTSTACK=2048
+CONFIG_ARCH_IRQBUTTONS=y
```
Ditto

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/Makefile:
```diff
@@ -0,0 +1,146 @@
+ifeq ($(CONFIG_ARCH_BUTTONS),y)
+CSRCS += sam_buttons.c
+endif
+
```
Please remove the above.  The Giant board has no buttons.

adamfeuer replied:
@patacongo will fix.

patacongo commented on boards/arm/sama5/giant-board/src/sam_ethernet.c:
```diff
@@ -0,0 +1,328 @@
+    {
+      sam_pioirqdisable(IRQ_INT_ETH0);
+    }
+}
```
The SAMD2 does not have a gigabit MAC, does it?  I think this needs to be cleaned up.  It has 2x100mbit MACs, right?

The naming is confusing because some Atmel documentation refers to the 100mbit MAC as a GMAC.  The code refers to these and EMAC.  I think you need to remove all references to GMAC unless I am not understanding.

adamfeuer replied:
@patacongo SAMA5D2 has a GMAC peripheral that is a single 10 or 100 Mbps Ethernet. I will look at the Atmel example code to see whether it is a GMAC or EMAC, then update the references.

adamfeuer replied:
@patacongo Since the board does not bring out the Ethernet pins, will remove the file.

patacongo commented on boards/arm/sama5/giant-board/src/sam_ethernet.c:
```diff
@@ -0,0 +1,328 @@
+#ifndef CONFIG_SAMA5_EMACA
+#  undef CONFIG_SAMA5_EMAC_ISETH0
+#endif
+
```
Remove, only SAMA5D3 has EMACA.

adamfeuer replied:
@patacongo will correct.

patacongo commented on boards/arm/sama5/giant-board/src/sam_ethernet.c:
```diff
@@ -0,0 +1,328 @@
+#ifdef CONFIG_SAMA5_EMAC_ISETH0
+#  define SAMA5_EMAC_DEVNAME "eth0"
+#  define SAMA5_GMAC_DEVNAME "eth1"
+#else
```
Remove all references to GMAC.  See below.

adamfeuer replied:
@patacongo Will correct, if the peripheral is an EMAC.

patacongo replied:
In NuttX, the 10/100mbit GMAC peripheral is referred to as EMAC.

That is because the same code is used for the entire family.  The SAMAD3 is the only family to have a gigabit GMAC and a 10/100 mbit EMAC.  The SAMAD2 and 4 and got peripherals called GMAC, but they are the same as the D3 EMAC, not the GMAC.  So in order to share the code, they have to be called EMAC.

But this is mute.  You need to remove all Ethernet support anyway.  The Giant Board does not bring other the Ethernet so you can't have any kind of Ethernet on that board (or NAND or buttons).

Let's not put unused and unusable files in the port.  If, in the future, someone hacks the Giant Board to support Ethernet, then we can bring the correct driver in (the one there now is not correct).

patacongo commented on boards/arm/sama5/giant-board/src/sam_i2schar.c:
```diff
@@ -0,0 +1,111 @@
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
```
You can remove empty section headers

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/sam_nandflash.c:
```diff
@@ -0,0 +1,212 @@
+/****************************************************************************
```
This should be removed.  There is NAND on the Giant Board.  It just confuses things cloning files which are not appropriate for the board.  It is a bad practice.  Please remove.

adamfeuer replied:
@patacongo will fix, sorry.

patacongo commented on boards/arm/sama5/giant-board/src/sam_ostest.c:
```diff
@@ -0,0 +1,99 @@
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
```
You can remove empty headers

adamfeuer replied:
@patacongo will fix

patacongo commented:
My general feeling is that this port is a little too immature to come into the repository.  The main problem is that a lot of things were clone apparently from SAMA5D4 and too many bad files and inappropriate logic came in.

it would be better to strip out all unnecessary and things that do not apply to the Giant board:  NAND support, Button support, Ethernet support.  That just degrades the quality of the board port.

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,349 @@
+#if !defined(CONFIG_NET) || (!defined(CONFIG_SAMA5_EMACA) && !defined(CONFIG_SAMA5_GMAC))
+#  undef HAVE_NETWORK
+#endif
+
```
Please remove the above.  The EMAC peripherals are not brought out for the Giant board.

adamfeuer replied:
@patacongo will fix if I don't need for Ethernet over USB.

patacongo replied:
You need to remove the ` (!defined(CONFIG_SAMA5_EMACA) && !defined(CONFIG_SAMA5_GMAC)`.  That is most certainly irrelevant and wrong.

adamfeuer replied:
@patacongo I did as you suggested and removed the EMACA and GMAC.

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,349 @@
+#define HAVE_USBHOST    1
+#define HAVE_USBDEV     1
+#define HAVE_USBMONITOR 1
+#define HAVE_NETWORK    1
```
Please remove.  The EMAC peripheral is not usable on the Giant Board.

adamfeuer replied:
@patacongo Ethernet over USB is definitely quite usable, and I have tested it, so I need to leave HAVE_NETWORK in. Let me know if that's not right–

patacongo replied:
If HAVE_ETHERNET is useful to you, then keep it as long as it has no dependency on the EMAC or GMAC peripheral which cannot be used.

adamfeuer replied:
@patacongo ok

patacongo commented on boards/arm/sama5/giant-board/src/sam_ethernet.c:
```diff
@@ -0,0 +1,328 @@
+/****************************************************************************
```
This should be removed.  The EMAC peripherals are not brought out on the Giant Board.  So this file is inappropriate and misleading.

adamfeuer replied:
@patacongo Same comment as above re: Ethernet over USB. I will look into exactly how that's implemented. Will remove if I don't need it for Ethernet over USB.

patacongo replied:
If you need anything releated to the non-function EMAC/GMAC peripheral then something is horribly screwed up in my implementation.  USB CDC/ECM absolutely does not need it and the perpherial is non-functional since the pins are not exposed.

patacongo replied:
You understand that the Ethernet MAC that is used resided the the USB device that is connected to SAMA5 via USB host.  That is the MAC used by the network, not the SAMA5 MAC which you could not use if you wanted to.

adamfeuer replied:
@patacongo Ok, will leave HAVE_ETHERNET and remove the other EMAC/GMAC references. Thank you for clarifying.

patacongo replied:
It is only used in sam_ethernet.c which should be removed anyway

adamfeuer replied:
@patacongo Ok, will search for references, and remove if I can.

patacongo commented on boards/arm/sama5/giant-board/configs/netnsh/defconfig:
```diff
@@ -0,0 +1,119 @@
+CONFIG_NSH_ARCHINIT=y
+CONFIG_NSH_BUILTIN_APPS=y
+CONFIG_NSH_FILEIOSIZE=512
+CONFIG_PREALLOC_TIMERS=4
```
Wow.  This is obviously old and out of date.  Pre-allocated timers disappear sometime back.. Please refresh.

patacongo commented on boards/arm/sama5/giant-board/configs/nsh/defconfig:
```diff
@@ -0,0 +1,80 @@
+CONFIG_NXPLAYER_DEFAULT_MEDIADIR="/mnt/sdcard"
+CONFIG_NXPLAYER_INCLUDE_SYSTEM_RESET=y
+CONFIG_NXPLAYER_RECURSIVE_MEDIA_SEARCH=y
+CONFIG_PREALLOC_TIMERS=4
```
??? Wasn't this option removed???

patacongo commented on boards/arm/sama5/giant-board/configs/sdmmcnsh/defconfig:
```diff
@@ -0,0 +1,143 @@
+CONFIG_NSH_PROMPT_STRING="nsh>  "
+CONFIG_NSH_QUOTE=y
+CONFIG_NSH_READLINE=y
+CONFIG_PREALLOC_TIMERS=4
```
???

patacongo commented:
Oops... Sorry.  Didn't mean to close the PR.

patacongo commented on boards/arm/sama5/giant-board/src/Makefile:
```diff
@@ -0,0 +1,146 @@
+
+ifeq ($(CONFIG_SAMA5_EBICS3_NAND),y)
+CSRCS += sam_nandflash.c
+endif
```
Please remove the above.  The Giant Board has no NAND

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,349 @@
+
+#define HAVE_SDMMC      1
+#define HAVE_AT25       1
+#define HAVE_NAND       1
```
Remvoe.  The Giant Board has no NAND

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,349 @@
+/* There is no NAND flash. */
+
+#undef HAVE_NAND
+
```
Remove.  The Giant Board has no NAND

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,349 @@
+/* There is no AT25 Serial flash */
+
+#undef HAVE_AT25
+#define _AT25_MINOR _NAND_MINOR
```
There is no NAND.  This needs to be fixed.

adamfeuer replied:
@patacongo will fix

adamfeuer commented on boards/arm/sama5/giant-board/configs/netnsh/defconfig:
```diff
@@ -0,0 +1,119 @@
+CONFIG_ARCH="arm"
+CONFIG_ARCH_BOARD="giant-board"
+CONFIG_ARCH_BOARD_GIANT_BOARD=y
+CONFIG_ARCH_BUTTONS=y
```
@patacongo I tested it it. As stated, I left the buttons code in thinking that the user might add their own; but will remove it.

adamfeuer commented on boards/arm/sama5/giant-board/src/Makefile:
```diff
@@ -0,0 +1,146 @@
+CSRCS += sam_ethernet.c
+endif
+endif
+
```
@patacongo On the SAMA5D2-XULT, I thought I needed sam_ethernet.c to use the CDC/ECM Ethernet over USB driver. I specifically need Ethernet over USB for my application. I will look into whether it's required. If I'm wrong about that, let me know!

adamfeuer replied:
@patacongo I removed this.

adamfeuer commented:
@patacongo It looks like a lot of your comments and suggestions would also apply to the SAMA5-XULT board files, where this code was ported from. When this PR is updated and accepted, would you also like to see a PR that applies similar fixes to the SAMA5-XULT board files?

patacongo commented:
> @patacongo <https://github.com/patacongo> Will correct, if the 
> peripheral is an EMAC.
>
As I mentioned in a different comment.  For SAMA5, all 10/100 mbit 
Ethernet is referred to as EMAC and never as GMAC.  Atmel refers to the 
10/100 mbit peripheral as EMAC on the SAMA5D3 but as GMAC on the SAMA5D2 
and SAMAD4.  These are all referred as EMAC for consistency in the 
configurations and code.  Only the SAMAD3 has a gigabit ethernet 
interface that it refers to as GMAC.

But this doesn't matter.  You need to remove all Ethernet support.  
Ethernet is not available on Giant Board (and no, it is not needed for 
CDC/ECM).

patacongo commented:
> @patacongo <https://github.com/patacongo> will fix if I don't need for 
> Ethernet over USB.
>
You do need Ethernet over USB to use CDC/ECM.  But you DO NOT need the 
SAMA5 Ethernet peripheral.  Please remove the driver for that.  Ethernet 
is handled by the USB CDC/ECM host and the NuttX network.  The SAMA5 
Ethernet MAC is not involved.

patacongo commented:
>
>> ------------------------------------------------------------------------
>>
>> In boards/arm/sama5/giant-board/src/sam_ethernet.c 
>> <https://github.com/apache/incubator-nuttx/pull/1728#discussion_r484529219>:
>>
>> > @@ -0,0 +1,328 @@
>> +/****************************************************************************
>>
>> @patacongo <https://github.com/patacongo> Ok, will leave 
>> HAVE_ETHERNET and remove the other EMAC/GMAC references. Thank you 
>> for clarifying.
>>
And what if you can't?  Then you have something really wrong that must 
be fixed before merging.  There is no involvement or relationship of the 
non-functional SAMA5 EMAC the USB CDC/ECM that uses the MAC in the 
connected USB device.  If there is some dependency it is wrong!

adamfeuer commented:
@patacongo re: HAVE_ETHERNET, yes, ok, will remove.

adamfeuer commented:
@patacongo By the way, I'm not as familiar with NuttX internals as you are, which is why I was qualifying my comments. I'm just trying to learn and do my best to code something that works and meets the NuttX project's standards.

patacongo commented:
How about CAN?  Are there really CAN interface brought out from the chip to accessible pins?  If not, we should remove the CAN driver as well.

patacongo commented:
> 
> 
> How about CAN? Are there really CAN interface brought out from the chip to accessible pins? If not, we should remove the CAN driver as well.

Bottom line:  We must not merge any untested code into the repository.  That would be shoddy engineering and a disgrace to the OS.  Please remove all of the files... including CAN... If you have not tested and cannot personally vouch for the correctness of the code, it MUST be remove.  It can come later when it is fully verified.  But not unverified code, please.

adamfeuer commented:
@patacongo Re: CAN, you are right, CANRX and CANTX are not on the pins, will remove, working on it. Sorry for the late response, was away from email this afternoon.

patacongo commented:
> 
> 
> @patacongo Re: CAN, you are right, CANRX and CANTX are not on the pins, will remove, working on it. Sorry for the late response, was away from email this afternoon.


Good!  I think it would also need a CAN transceiver and connector to be usable.  I think with this board we would need to used something like an MCP 2515.

I think we are on the way to a clean BSP for the Giant Board that we can be proud of.  I have started thinking about buying one again (you can get them from Mouser now).  You have inspired me.  We should announce the support on Twitter... Groguard is there as well as other members of the Giant Board community.

adamfeuer commented:
@patacongo @ghn-certi I think I fixed all the issues you commented on. And as requested I removed all the files for non-included features (Ethernet, NAND flash, SPI flash, etc.). Will you take a look and see if I missed anything, or if there's anything else that needs to be changed?

adamfeuer commented:
The build is failing, but I don't see any errors in the arm-05 logs. Not sure what to do here.

btashton commented:
> The build is failing, but I don't see any errors in the arm-05 logs. Not sure what to do here.

@adamfeuer you need to make sure you are rebased and then refresh your configs.  You probably need to do it on all your configs, in the logs you will see entries like this:
```
====================================================================================
Configuration/Tool: giant-board/sdmmc-net-nsh,CONFIG_ARMV7M_TOOLCHAIN_GNU_EABIL
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Disabling CONFIG_ARMV7A_TOOLCHAIN_GNU_EABIL
  Enabling CONFIG_ARMV7M_TOOLCHAIN_GNU_EABIL
  Building NuttX...
  Normalize giant-board/sdmmc-net-nsh
80d79
< CONFIG_NFILE_STREAMS=8
Saving the new configuration file
HEAD detached at pull/1728/merge
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   boards/arm/sama5/giant-board/configs/sdmmc-net-nsh/defconfig
```

adamfeuer commented:
@btashton Got it. I will do that. BTW what exactly does "normalize" do? I looked briefly at the CI code, but didn't see what it was doing there. Ideally I would run that locally before submitting.

adamfeuer commented:
@btashton Ok, I see the CI script runs `testbuild.sh`, I'll try that locally.

adamfeuer commented:
@patacongo @ghn-certi Ok,as noted above, I think I corrected the issues you raised, and I refreshed the configs so that the tests pass. The build is failing on some nxstyle changes that I don't see an easy way to fix. 

Does this look ok to you now?

v01d commented:
The CI is strange, it is acting on a commit range not from this PR. The list of commits is bigger.
You could try to re-run the workflow or rebase and push to see if this fixes it.

adamfeuer commented:
Thanks @v01d, I rebased+squashed and force-pushed. Let's see how it does now.

patacongo commented:
> 
> 
> @patacongo @ghn-certi Ok,as noted above, I think I corrected the issues you raised, and I refreshed the configs so that the tests pass. The build is failing on some nxstyle changes that I don't see an easy way to fix.
> 
> Does this look ok to you now?

I'll take a look in the morning.

v01d commented:
Ok, now it worked :shrug: 
Some easy to fix style problems only

patacongo commented on boards/Kconfig:
```diff
@@ -2268,7 +2277,8 @@ config ARCH_BOARD
 	default "sabre-6quad"              if ARCH_BOARD_SABRE_6QUAD
-	default "sama5d2-xult"             if ARCH_BOARD_SAMA5D2_XULT
+    default "sama5d2-xult"             if ARCH_BOARD_SAMA5D2_XULT
+    default "giant-board"              if ARCH_BOARD_GIANT_BOARD
```
Looks like spaces were used here instead of TABs.  Can you fix this?

adamfeuer replied:
@patacongo Oops sorry, will fix. Have to change my editor settings for that! 🙂

patacongo commented on boards/arm/sama5/giant-board/Kconfig:
```diff
@@ -0,0 +1,127 @@
+    ---help---
+    Enable SD Card interface SDMMC0. Selects SAMA5_SDMMC SAMA5_SDMMC0
+    SAMA5_SDMMC_DMA SDIO_DMA SCHED_WORKQUEUE SCHED_HPWORK SDIO_BLOCKSETUP
+
```
In most usage, help text is indented by two TABs.

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/Kconfig:
```diff
@@ -0,0 +1,127 @@
+    default 4294967296
+    depends on SAMA5_SDMMC0
+    ---help---
+    Size of eMMC flash in bytes. Default: 4GB
```
2 TABs

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/Kconfig:
```diff
@@ -0,0 +1,127 @@
+    default 1073741824
+    depends on SAMA5_SDMMC1
+    ---help---
+    Size of SD Card in bytes. Default: 16GB
```
2 TABs

adamfeuer replied:
@patacongo will fix

patacongo commented on boards/arm/sama5/giant-board/src/Makefile:
```diff
@@ -0,0 +1,105 @@
+
+ifeq ($(CONFIG_SAMA5_DDRCS),y)
+CSRCS += sam_sdram.c
+endif
```
Remove.  sam_dram.c does not exist, does it?

patacongo commented on boards/arm/sama5/giant-board/src/Makefile:
```diff
@@ -0,0 +1,105 @@
+CSRCS += sam_i2schar.c
+endif
+endif
+endif
```
Does sam_i2schar.c exist?  Think not.  These should be removed.

patacongo replied:
Is I2S available on the Giant Board?  I think not but I would need to check.

patacongo commented on boards/arm/sama5/giant-board/include/board_492mhz.h:
```diff
@@ -0,0 +1,207 @@
+ */
+
+#define SDMMC_INIT_CLKDIV          (164 << SDMMC_MR_CLKDIV_SHIFT)
+
```
HSMCI is not available on the Giant board, is it?  I would need to check.  The SD/MMC interface is via SPI, isn't it?

patacongo replied:
In the SAMA5D2, HSMCI is referred as SDMMC, isn't it?

patacongo commented on boards/arm/sama5/giant-board/include/board_sdram.h:
```diff
@@ -0,0 +1,167 @@
+#define HSMCI_SDXFR_CLKDIV         sam_hsmci_clkdiv(25000000)
+
+#define HSMCI_SDWIDEXFR_CLKDIV     HSMCI_SDXFR_CLKDIV
+
```
HSMCI is not available on the Giant Board, is it?  Shouldn't these be removed?

patacongo replied:
In the SAMA5D2, HSMCI is referred as SDMMC, isn't it?

patacongo commented on boards/arm/sama5/giant-board/src/Makefile:
```diff
@@ -0,0 +1,105 @@
+CSRCS += sam_hsmci.c
+endif
+endif
+
```
sam_hsmci.c does not exist, does it?  I don't think that the HSMCI interface is available on the Giant Board.  If HSMCI is not available, any hsmci files should be removed.

patacongo replied:
In the SAMA5D2, HSMCI is referred as SDMMC, isn't it?

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,436 @@
+/* Configuration ************************************************************/
+
+#define HAVE_SDMMC      1
+#define HAVE_USBHOST    1
```
There is not SDMMC (i.e., HSCMI).  SD card support is via SPI.

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,436 @@
+#if !defined(CONFIG_SAMA5_SDMMC) && !defined(CONFIG_SAMA5_SDMMC0)
+#  undef HAVE_SDMMC
+#endif
+
```
Delete, SDMMC is not support on the Giant Board.

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,436 @@
+#  endif
+#else
+#endif
+
```
Remove all of the above.  The SDMMC interface is not available on the Giant Board, is it?

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,436 @@
+ * protect discrete is not connected and not available to software.  The
+ * slot only supports 4-bit wide transfer mode, and the NuttX driver
+ * currently uses only the 4-bit wide transfer mode.
+ *
```
These comments are incorrect.  The SD card connects via SPI, correct?

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,436 @@
+#ifdef HAVE_SDMMC
+int sam_sdmmc_initialize(int slotno, int minor);
+#endif
+
```
This should be removed.  The the SDMMC (HSMCI) interface is not supported on the Giant Board.

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,400 @@
+#  include <nuttx/mmcsd.h>
+#  include "sam_sdmmc.h"
+#endif
+
```
The Giant Board does not support the SDMMC interface.

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,400 @@
+#ifdef CONFIG_SAMA5_SDMMC
+
+static int nsh_sdmmc_initialize(void)
+{
```
Remove this function.  The Giant Board does not support the SDMMC interface.

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,400 @@
+  nsh_sdmmc_initialize();
+#endif
+#endif
+
```
Remove the above SDMMC not supported.

patacongo commented on boards/arm/sama5/giant-board/src/sam_bringup.c:
```diff
@@ -0,0 +1,400 @@
+      auderr("ERROR: Failed to initialize WM8904 audio: %d\n", ret);
+    }
+#endif
+
```
R

patacongo commented on boards/arm/sama5/giant-board/configs/sdmmc-net-nsh/defconfig:
```diff
@@ -0,0 +1,151 @@
+# You can then do "make savedefconfig" to generate a new defconfig file that includes your
+# modifications.
+#
+CONFIG_ARCH="arm"
```
This configuration should be removed since it depends on SDMMC which is not available.

Please, save us a lot of problems in the future:  Do not commit any defconfig files that you have not tested.

patacongo commented on boards/arm/sama5/giant-board/include/board.h:
```diff
@@ -0,0 +1,195 @@
+
+#define BOARD_SDMMC_SD4MODE_PRESCALER   SDMMC_SYSCTL_SDCLKFS_DIV8
+#define BOARD_SDMMC_SD4MODE_DIVISOR     SDMMC_SYSCTL_DVS_DIV(1)
+
```
SDIO is STM32 terminology for HSMCI/SDMMC.  These should be removed since SDMMC is not supported by the Giant Board.

patacongo commented on boards/arm/sama5/giant-board/src/giant-board.h:
```diff
@@ -0,0 +1,436 @@
+
+#define SDMMC_SDXFR_CLKDIV         (1 << SDMMC_MR_CLKDIV_SHIFT)
+#define SDMMC_SDWIDEXFR_CLKDIV     SDMMC_SDXFR_CLKDIV
+
```
Remove the above.  The Giant Board does support the SDMMC interface.

patacongo commented on boards/arm/sama5/giant-board/src/sam_i2schar.c:
```diff
@@ -0,0 +1,111 @@
+      ret = i2schar_register(i2s, CONFIG_GIANT_BOARD_I2SCHAR_MINOR);
+      if (ret < 0)
+        {
+          aerr("ERROR: i2schar_register failed: %d\n", ret);
```
```suggestion
          auderr("ERROR: i2schar_register failed: %d\n", ret);
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_i2schar.c:
```diff
@@ -0,0 +1,111 @@
+      i2s = sam_ssc_initialize(CONFIG_GIANT_BOARD_SSC_PORT);
+      if (!i2s)
+        {
+          _err("ERROR: Failed to get the SAMA5 SSC/I2S driver for SSC%d\n",
```
```suggestion
          auderr("ERROR: Failed to get the SAMA5 SSC/I2S driver for SSC%d\n",
```

patacongo commented on boards/arm/sama5/giant-board/src/sam_sdmmc.c:
```diff
@@ -0,0 +1,319 @@
+ *   PA28  SDMMC1_CMD
+ *
+ */
+
```
I believe that all of the comments above are incorrect.

patacongo commented on boards/arm/sama5/giant-board/src/sam_sdmmc.c:
```diff
@@ -0,0 +1,319 @@
+/****************************************************************************
+ *  boards/arm/sama5/giant-board/src/sam_sdmmc.c
+ *
```
This entire file must be removed.  The Giant Board does not support the SDMMC (HSMCI) interface.  The interface to the SD card has to be via SPI.

patacongo commented on boards/arm/sama5/giant-board/src/sam_sdram.c:
```diff
@@ -0,0 +1,572 @@
+  /* SAMA5D27D1G uses Winbond W971GG6SB 1GB DDR2 RAM
+   *
+   * See datasheet http://ww1.microchip.com/downloads/en/DeviceDoc/SAMA5D2-Plus-DDR2-LPDDR2-System-in-Package-(SIP)-60001484b.pdf
+   *
```
it is OK to break URLs into shorten segments.  It is better to remove the long line error than to try to preserve the entire URL as a single string.

adamfeuer replied:
@patacongo Got it, will fix.

patacongo commented:
@adamfeuer Thanks for the update.  I have a a few more small comments.  I thought for a awhile that the SDMMC is not available on the Giant Board because the pins are not brought out.  Then I noticed that the SDMMC-based card slot is on the back of the card.  I think I backed out all of those comments.  Sorry about that.

patacongo commented on boards/arm/sama5/giant-board/src/sam_sdram.c:
```diff
@@ -0,0 +1,572 @@
+
+  /* SAMA5D27D1G uses Winbond W971GG6SB 1GB DDR2 RAM
+   *
+   * See datasheet http://ww1.microchip.com/downloads/en/DeviceDoc/SAMA5D2-Plus-DDR2-LPDDR2-System-in-Package-(SIP)-60001484b.pdf
```
```suggestion
   * See datasheet http://ww1.microchip.com/downloads/en/DeviceDoc/
   *     SAMA5D2-Plus-DDR2-LPDDR2-System-in-Package-(SIP)-60001484b.pdf
```

adamfeuer commented:
@patacongo Ok, I took care of the tabs/spaces stuff and accepted your suggestions. Is there more to do? 

If not, let me know, and I will rebase and force-push to get ready for merging.

patacongo reviewed (approved):
LGTM.

We can ignore the nxstyle complaints:

    ##[warning]/home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/boards/arm/sama5/giant-board/include/board.h:63:3: warning: #include outside of 'Included Files' section
    ##[warning]/home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/boards/arm/sama5/giant-board/include/board.h:72:3: warning: #include outside of 'Included Files' section

patacongo commented:
I will go ahead an merge this.  It looks like there is breakage from some other source now.  There are STM32 compilations failing.  This PR past the checks before and could not be causing STM32 failures.

patacongo commented:
One thing that I missed in my review is that there is also documentation that needs to be updated for the new board support:

- README.md
- boards/README.txt
- Documentation/introduction/detailed_support.rst

Could you update these too?  Should take just a moment.

btashton commented:
@patacongo @adamfeuer This broke master.  I just checked out the PR and it also fails because of issues in the Kconfig

```
boards/arm/sama5/giant-board/Kconfig:110: syntax error
boards/arm/sama5/giant-board/Kconfig:109: invalid option
make: *** [tools/Makefile.unix:464: olddefconfig] Error 1
ERROR: failed to refresh
```

Can one of you fix this or should I dig into it?

btashton commented:
This breakage was minor see #1807 for fix.  To be clear the CI failure in the PR was due to this, it just was confusing because all arch were failing because config failed but the build pushed on to the build phase with an invalid configuration

adamfeuer commented:
@btashton Thanks for fixing. I'm sorry I missed that syntax error— I should have run a config refresh.

adamfeuer commented:
@patacongo There is an updated README.md for the Giant Board. I will submit another PR with additions for the boards/README.txt.

patacongo commented:
> 
> 
> @patacongo There is an updated README.md for the Giant Board. I will submit another PR with additions for the boards/README.txt.

I was referring to the top-level nuttx/README.md (and alsoDocumentation/introduction/detailed_support.rst), both of which have full lists of links to ALL README files.

And, as you note, boards/README.txt.

adamfeuer commented:
@patacongo See #1809 for updates to READMES and docs. I didn't see any updates needed for the top-level README.md... if I do need to add something to this, will you let me know?

patacongo commented:
> 
> 
> @patacongo See #1809 for updates to READMES and docs. I didn't see any updates needed for the top-level README.md... if I do need to add something to this, will you let me know?

This should have been updated in the top-level README:

     |   |   |- sama5/
     |   |   |   |- sama5d2-xult/
     |   |   |   |   `- README.txt
     |   |   |   |- sama5d3x-ek/
     |   |   |   |   `- README.txt
     |   |   |   |- sama5d3-xplained/
     |   |   |   |   `- README.txt
     |   |   |   `- sama5d4-ek/
     |   |   |       `- README.txt

adamfeuer commented:
@patacongo Thanks for catching, see #1811 for the fix. :slightly_smiling_face:
